components/style/values/specified/box.rs

```diff
@@ -492,3 +493,86 @@ pub fn assert_touch_action_matches() {
         NS_STYLE_TOUCH_ACTION_MANIPULATION => TouchAction::TOUCH_ACTION_MANIPULATION,
     }
 }
+
+bitflags! {
+    #[derive(MallocSizeOf, ToComputedValue)]
+    /// Constants for contain: https://drafts.csswg.org/css-contain/#contain-property
+    pub struct Contain: u8 {
+        /// `layout` variant, turns on layout containment
+        const LAYOUT = 0x01;
+        /// `style` variant, turns on style containment
+        const STYLE = 0x02;
+        /// `paint` variant, turns on paint containment
+        const PAINT = 0x04;
+        /// `strict` variant, turns on all types of containment
+        const STRICT = 0x8;
+        /// `strict bits` variant
+        const STRICT_BITS = Contain::LAYOUT.bits | Contain::STYLE.bits | Contain::PAINT.bits;
+    }
+}
+
+impl ToCss for Contain {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        if self.is_empty() {
+            return dest.write_str("none")
+        }
+        if self.contains(Contain::STRICT) {
+            return dest.write_str("strict")
+        }
+
+        let mut has_any = false;
+        macro_rules! maybe_write_value {
+            ($ident:path => $str:expr) => {
+                if self.contains($ident) {
+                    if has_any {
+                        dest.write_str(" ")?;
+                    }
+                    has_any = true;
+                    dest.write_str($str)?;
+                }
+            }
+        }
+        maybe_write_value!(Contain::LAYOUT => "layout");
+        maybe_write_value!(Contain::STYLE => "style");
+        maybe_write_value!(Contain::PAINT => "paint");
+
+        debug_assert!(has_any);
+        Ok(())
+    }
+}
+
+impl Parse for Contain {
+    /// none | strict | content | [ size || layout || style || paint ]
+    fn parse<'i, 't>(_context: &ParserContext, input: &mut Parser<'i, 't>)
+                         -> Result<Contain, ParseError<'i>> {
+        let mut result = Contain::empty();
```

Comment:
This function is kinda dumb, and could be rewritten to be faster / clearer. Can you file a followup for it?

In particular, I'm thinking of something like:

```rust
let mut result = Contain::empty();
while let Ok(name) = input.try(|i| i.expect_ident_cloned()) {
    match_ignore_ascii_case! { &name,
        "layout" => Some(Contain::LAYOUT),
        "style" => Some(Contain::STYLE),
        "paint" => Some(Contain::PAINT),
        // The following are only valid as the first value, thus the is_empty()
        // checks.
        "strict" => {
            if result.is_empty() {
                return Ok(Contain::STRICT | Contain::STRICT_BITS)
            }
            None
        }
        "none" => {
            if result.is_empty() {
                return Ok(result);
            }
            None
        }
        _ => None
    };

    // Same code as now.
}
```

Also fine to do it in a separate commit in this same PR.

That avoids having to tokenize twice for `none` and `strict` just to tokenize an identifier again below.

Reply:
Thanks for the feedback, I will file a followup commit.